src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
     /// <summary>
     /// A recoverable TextAndVersion source that saves its text to temporary storage.
     /// </summary>
-    internal class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVersionable
+    internal sealed class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVersionable
     {
-        private readonly ITemporaryStorageServiceInternal _storageService;
+        private readonly HostWorkspaceServices _services;
 
-        private SemaphoreSlim? _lazyGate;
-        private ValueSource<TextAndVersion>? _initialSource;
+        // ValueSource<TextAndVersion> or RecoverableText
+        private object _initialSourceOrRecoverableText;
 
-        private RecoverableText? _text;
-        private VersionStamp _version;
-        private string? _filePath;
-        private Diagnostic? _loadDiagnostic;
-
-        public RecoverableTextAndVersion(
-            ValueSource<TextAndVersion> initialTextAndVersion,
-            ITemporaryStorageServiceInternal storageService)
+        public RecoverableTextAndVersion(ValueSource<TextAndVersion> initialSource, HostWorkspaceServices services)
         {
-            _initialSource = initialTextAndVersion;
-            _storageService = storageService;
+            _initialSourceOrRecoverableText = initialSource;
+            _services = services;
         }
 
-        private SemaphoreSlim Gate => LazyInitialization.EnsureInitialized(ref _lazyGate, SemaphoreSlimFactory.Instance);
-
-        public ITemporaryTextStorageInternal? Storage => _text?.Storage;
+        public ITemporaryTextStorageInternal? Storage
+            => (_initialSourceOrRecoverableText as RecoverableText)?.Storage;
 
         public override bool TryGetValue([MaybeNullWhen(false)] out TextAndVersion value)
         {
-            if (_text != null && _text.TryGetValue(out var text))
+            var sourceOrRecoverableText = _initialSourceOrRecoverableText;
+
+            if (sourceOrRecoverableText is RecoverableText recoverableText)
             {
-                value = TextAndVersion.Create(text, _version, _filePath, _loadDiagnostic);
-                return true;
+                if (recoverableText.TryGetValue(out var text))
+                {
+                    value = TextAndVersion.Create(text, recoverableText.Version, filePath: null, recoverableText.LoadDiagnostic);
+                    return true;
+                }
             }
-            else
+            else if (((ValueSource<TextAndVersion>)sourceOrRecoverableText).TryGetValue(out value))
```

Comment:
```suggestion
            else
            {
                return ((ValueSource<TextAndVersion>)sourceOrRecoverableText).TryGetValue(out value);
            }
```

Reply:
there's a race condition in this code (and likely elsewhere).  You could check `if (sourceOrRecoverableText is RecoverableText recoverableText)` and have it be false.  Then, prior to getting to this else block, another thread swaps sourceOrRecoverableText out for a REcoverableText.  Now thsi cast will crash.  

Can we just switch to a nice fat lock (or semaphoreslim)? :)

Reply:
ok.  ignore this comment.  i see that you made a local :)  can you doc that you made a local?